docs/proposals/PROPOSAL_B.md

```diff
@@ -0,0 +1,287 @@
+* Namespace - a validation area under which users can describe plugin specs that are consumed by plugins (executable binaries) as input.
+* Plugin - an executable binary that extracts data and checks compatibility on a host.
+
+A new imageCompatibility artifact type that contains a layer with a JSON compatibility object.
+The object represents a graph of namespaces.
+Namespace describes the validation area and is another object containing data that provides plugins specification for the corresponding plugins.
+A plugin is an executable binary that validates the host and returns information about success or failure.
+
+### Compatibility Spec
+
+Compatibility spec is a graph that can have multiple roots with different paths.
+If one namespace in the path fails, the path is incompatible with the host.
+Namespace fails if at least one plugin reports an error.
+If all paths fail, the host is not compatible with the container.
+
+The host is compatible with the container if at least one path is completed successfully.
+
+#### Schema
+
+- **`schemaVersion`** *int*
+
+  This REQUIRED property specifies the image manifest schema version. For this version of the specification.
+
+- **`namespaces`** *string-object map*
+
+  This REQUIRED property specifies a map of validation areas.
+
+  - **`<namespace-name>`** *object*
+
+    This REQUIRED property specifies the validation area.
+
+    - **`plugins`** *string-object*
+
+      This OPTIONAL property is a list of plugins specifications.
+
+        - **`<plugin-name>`** *object*
+
+          This REQUIRED property specifies the plugin specification for plugin executable binary that validates the host and returns information about success or failure
+
+          - **`source`** *string*
+
+            This OPTIONAL property specifies an HTTP address of the plugin.
+
+          - **`data`** *object*
+
+            This OPTIONAL property specifies the input data for plugin executable binary. 
+
+    - **`refs`** *object*
+
+      This OPTIONAL property specifies the references to other namespaces.
+
+      - **`all`** *array of strings*
+
+        This OPTIONAL property specifies a list of all required namespaces that have to pass that container is considered to be compatible with a host.
+
+      - **`oneOf`** *array of strings*
+
+        This OPTIONAL property specifies a list of namespaces with a condition that at least one namespace has to pass that container is considered to be compatible with a host.
+
+      - **`noneOf`** *array of strings* 
+
+        This OPTIONAL property specifies a list of namespaces that cannot pass that container is considered to be compatible with a host.
+
+##### Example
+
+<details>
+  <summary>GPU pass-through</summary>
+
+```json
+{
+  "schema": "0.1.0",
+  "namespaces": {
+    "kernelMainlineV5": {
+      "plugins": {
+        "kernelVersion": {
+          "source": "<http-something-to-download-plugin>",
+          "data": {
+            "range": "5.0.0-5.15.0"
```

Comment:
I remembered that we discussed around the scope representation, it's better to use the math interval, so that whether to including the start/end in the scope, for example, ["5.0.0", "5.15.0"), means greater or equal to "5.0.0", and less than "5.15.0".

Reply:
Use space ABI in Linux is by default really well backwards compatible, so asking for version numbers is fine, but is it due to a specific feature in a specific kernel version? If one is asking for BPF kernel features, one is removing a wide range of older kernels in the same go. And whose kernel was this (I remember Bjorn asking some time ago).

Reply:
for kernel version requirements, there are lots of actual use cases mentioned in this document section : https://docs.google.com/document/d/1eGlZOyJIp3ZLIcwblCK4WgfHskiOqAjpovWtDw9mMGU/edit#heading=h.12j444qyywg1,

other factors are as following:
* kernel version range are optional compatibility requirements, the kernel version could be upstream or downstream distribution.
* some features (for example write protected hugepage)cannot be described by kernel modules or configurations or boot args, have to use kernel version to make sure these features provided there.
* mission critical appplications will not promise to run on any kernel before testing, it's important for these mission critical applications to describe accurate  compatible kernel version or version range.
* not all apps are publicly visble but they are also running as containers, especially for mission critical apps. These apps will not be published to (and visible in) public image registries, therefore such kind of kernel requirements can rarely be found for general application in public registries.